README.md
# code-review-leaderboard

[![Build Status](https://dev.azure.com/lachiejames/code-review-leaderboard/_apis/build/status/lachiejames.code-review-leaderboard?branchName=main)](https://dev.azure.com/lachiejames/code-review-leaderboard/_build/latest?definitionId=12&branchName=main) [![codecov](https://codecov.io/gh/lachiejames/code-review-leaderboard/branch/main/graph/badge.svg?token=kNLA2ldKKF)](https://codecov.io/gh/lachiejames/code-review-leaderboard) [![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

**code-review-leaderboard** scrapes data from all pull requests across every repository within your organisation. This data is processed and consolidated into a leaderboard. Find out who is doing the most code reviews at your organisation 🥇

**Currently supports the following platforms:**

-   ✅ **Azure Repos**
-   ✅ **Github**
-   ✅ **Gitlab**

## Running from the command line

### Installing

Install `code-review-leaderboard` globally using npm:

```
npm install -g code-review-leaderboard
```

### Setting up your configuration and running locally

You can set up your configuration from the command line using:

```
code-review-leaderboard
```

It will then fetch the pull request data from the selected organisations, using the provided credentials. If successful, it will print out a leaderboard that looks like this:

![Example results](https://github.com/lachiejames/media-host/blob/main/code-review-leaderboard/example-results.png?raw=true)

### Demo

![Video demo](https://github.com/lachiejames/media-host/blob/main/code-review-leaderboard/demo.gif?raw=true)

## Running in TypeScript

### Installing

Install the project using:

```
git clone https://github.com/lachiejames/code-review-leaderboard.git
```

### Setting up your configuration

Open `code-review-leaderboard.config.js` from the root directory, and populate the following options:

| Property                   | Description                                     | Example                           |
| -------------------------- | ----------------------------------------------- | --------------------------------- |
| startDate                  | Include pull requests updated after this date   | Date("2021-05-01")                |
| endDate                    | Include pull requests updated before this date  | Date("2021-05-07")                |
| azure.enabled              | Include pull requests from Azure                | true                              |
| azure.baseUrl              | The Azure home page for your organisation       | `"https://dev.azure.com/Example"` |
| azure.personalAccessToken  | Authenticates HTTP requests to Azure            | "3Ccz4G6QPilk"                    |
| github.enabled             | Include pull requests from Github               | true                              |
| github.baseUrl             | The Github home page for your organisation      | `"https://github.com/Example"`    |
| github.personalAccessToken | Authenticates HTTP requests to Github           | "3Ccz4G6QPilk"                    |
| gitlab.enabled             | Include pull requests from Gitlab               | false                             |
| gitlab.baseUrl             | The Gitlab home page for your organisation      | `"https://gitlab.example.com/"`   |
| gitlab.personalAccessToken | Authenticates HTTP requests to Gitlab           | "Hf4sXcfn7M69"                    |
| httpTimeoutInMS            | Max time to wait for a HTTP request to complete | 5000                              |

### Running locally

Installing dependencies:

```
yarn
```

Compiling to JavaScript:

```
yarn build
```

Running with Node:

```
yarn start
```

Running tests:

```
yarn test
```

### More info

Check out this blog post that describes `code-review-leaderboard` in further detail:
https://lachiejames.com/spice-up-code-reviews-at-your-organisation-with-a-leaderboard/


CHANGELOG.md
## [1.3.1](https://github.com/lachiejames/code-review-leaderboard/compare/v1.3.0...v1.3.1) (2022-05-03)


### Bug Fixes

* remove data from config ([7e943f6](https://github.com/lachiejames/code-review-leaderboard/commit/7e943f6a6be632714051fa8e765948b839cb7940))

# [1.3.0](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.17...v1.3.0) (2022-05-03)


### Features

* add Github support ([ba18865](https://github.com/lachiejames/code-review-leaderboard/commit/ba188655f2aee2181324f8c5627a2537e0398a02))

## [1.2.17](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.16...v1.2.17) (2022-02-02)


### Bug Fixes

* remove tsconfig rule that broke build ([9fd29c9](https://github.com/lachiejames/code-review-leaderboard/commit/9fd29c9f057253fdb4cffd226cc3c08fa5b4c2e6))

## [1.2.16](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.15...v1.2.16) (2021-12-23)


### Bug Fixes

* **deps:** update dependency date-fns to v2.27.0 ([#99](https://github.com/lachiejames/code-review-leaderboard/issues/99)) ([750f824](https://github.com/lachiejames/code-review-leaderboard/commit/750f824548fd1dd22792b4bc687e1ce1527d6ad7))
* **deps:** update dependency ora to v6 ([#57](https://github.com/lachiejames/code-review-leaderboard/issues/57)) ([1eff1bf](https://github.com/lachiejames/code-review-leaderboard/commit/1eff1bfff9686b22247e0c2b26edb45587a14f3a))
* **deps:** update dependency table to v6.7.5 ([#98](https://github.com/lachiejames/code-review-leaderboard/issues/98)) ([dd4b21b](https://github.com/lachiejames/code-review-leaderboard/commit/dd4b21bb5c359e89e84b18138413f1118d0fc049))


### Reverts

* Revert "fix(deps): update dependency ora to v6 (#57)" (#101) ([fe1a11a](https://github.com/lachiejames/code-review-leaderboard/commit/fe1a11ad78c4cae4b618f60c72b4f12fb957e96a)), closes [#57](https://github.com/lachiejames/code-review-leaderboard/issues/57) [#101](https://github.com/lachiejames/code-review-leaderboard/issues/101)

## [1.2.15](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.14...v1.2.15) (2021-10-14)


### Bug Fixes

* remove lambda steps from pipeline ([6352c2f](https://github.com/lachiejames/code-review-leaderboard/commit/6352c2fe003d4db4e19d451489eb97ef83e53d70))

## [1.2.14](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.13...v1.2.14) (2021-10-14)


### Bug Fixes

* **deps:** update dependency date-fns to v2.25.0 ([#88](https://github.com/lachiejames/code-review-leaderboard/issues/88)) ([46706e9](https://github.com/lachiejames/code-review-leaderboard/commit/46706e95d3cb47e893d92122a868c3b20b527aec))
* **deps:** update dependency prompts to v2.4.2 ([#89](https://github.com/lachiejames/code-review-leaderboard/issues/89)) ([ad3e712](https://github.com/lachiejames/code-review-leaderboard/commit/ad3e7120f74c7d69ca6df7c92b3eaa75fd81d00c))
* **deps:** update dependency table to v6.7.2 ([#87](https://github.com/lachiejames/code-review-leaderboard/issues/87)) ([8ed716e](https://github.com/lachiejames/code-review-leaderboard/commit/8ed716eed0f36958c02d7e8b079eb55644b3ad38))

## [1.2.13](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.12...v1.2.13) (2021-09-19)


### Bug Fixes

* resolve lint errors (part 2) ([ab27d09](https://github.com/lachiejames/code-review-leaderboard/commit/ab27d09a2c03b192d40a265cf7d3b43895fad3b2))

## [1.2.12](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.11...v1.2.12) (2021-09-19)


### Bug Fixes

* resolve lint errors ([78cbc02](https://github.com/lachiejames/code-review-leaderboard/commit/78cbc0288ec8e39b98ef62df15950c71192a5e0e))

## [1.2.11](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.10...v1.2.11) (2021-09-19)


### Bug Fixes

* refine dev dependencies ([#80](https://github.com/lachiejames/code-review-leaderboard/issues/80)) ([48c5b28](https://github.com/lachiejames/code-review-leaderboard/commit/48c5b28eac39a1ea14ca3f4c48791bb7da5002df))

## [1.2.10](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.9...v1.2.10) (2021-09-18)


### Bug Fixes

* **deps:** update dependency date-fns to v2.24.0 ([#78](https://github.com/lachiejames/code-review-leaderboard/issues/78)) ([14c2b22](https://github.com/lachiejames/code-review-leaderboard/commit/14c2b220210b9e0caaba6953b622b568739c2d88))
* **deps:** update dependency gaxios to v4.3.2 ([#74](https://github.com/lachiejames/code-review-leaderboard/issues/74)) ([82b32fb](https://github.com/lachiejames/code-review-leaderboard/commit/82b32fb1f650c2edba6974dcf10bdea9039f20df))

## [1.2.9](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.8...v1.2.9) (2021-09-12)


### Bug Fixes

* **deps:** update dependency gaxios to v4.3.1 ([#65](https://github.com/lachiejames/code-review-leaderboard/issues/65)) ([e7f80c6](https://github.com/lachiejames/code-review-leaderboard/commit/e7f80c65c7c760953fabca0174bd4baa9aa2fda6))

## [1.2.8](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.7...v1.2.8) (2021-08-09)


### Bug Fixes

* Update README ([#49](https://github.com/lachiejames/code-review-leaderboard/issues/49)) ([6e3bbca](https://github.com/lachiejames/code-review-leaderboard/commit/6e3bbca3da66c45cab30ffbd61e0157c61aaa2f3))

## [1.2.7](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.6...v1.2.7) (2021-07-31)


### Bug Fixes

* **deps:** update dependency date-fns to v2.23.0 ([#45](https://github.com/lachiejames/code-review-leaderboard/issues/45)) ([4654788](https://github.com/lachiejames/code-review-leaderboard/commit/465478883ee5ce584fd77caae75b8555f427b309))

## [1.2.6](https://github.com/lachiejames/code-review-leaderboard/compare/v1.2.5...v1.2.6) (2021-06-14)


### Bug Fixes

* Reduce package size ([#23](https://github.com/lachiejames/code-review-leaderboard/issues/23)) ([667aece](https://github.com/lachiejames/code-review-leaderboard/commit/667aece2c0618718176640b07a2d2a5bf27e8260))


src/cli/prompts.ts
import { addDays } from "date-fns";
import prompts, { Answers } from "prompts";

import { Config } from "../shared/config.model";

import { validateEndDate, validatePersonalAccessToken, validateUrl } from "./validators";

// `prompts` package uses generics which accepts string literal values and then uses those values as types
// This means I have to ensure I use a string literal type that matches a string literal value
const PROMPT_NAME = "value";
type PromptType = "value";

export const getStartDate = async (): Promise<Date> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "date",
        message: "Choose a start date: ",
        initial: new Date(Date.now()),
        mask: "DD-MM-YYYY",
    });

    return promptData[PROMPT_NAME];
};

export const getEndDate = async (startDate: Date): Promise<Date> => {
    const datePlus2Weeks = addDays(new Date(startDate), 14);

    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "date",
        message: "Choose an end date: ",
        initial: datePlus2Weeks,
        mask: "DD-MM-YYYY",
        validate: (endDate: Date) => validateEndDate(startDate, endDate),
    });

    return promptData[PROMPT_NAME];
};

export const getPlatforms = async (): Promise<string[]> => {
    const promptData: Answers<PromptType> = await prompts({
        type: "multiselect",
        name: PROMPT_NAME,
        message: "Select your platforms: ",
        choices: [
            { title: "Azure", value: "Azure", selected: false },
            { title: "Github", value: "Github", selected: true },
            { title: "Gitlab", value: "Gitlab", selected: false },
        ],
        instructions: false,
        min: 1,
    });

    return promptData[PROMPT_NAME];
};

export const getAzureBaseUrl = async (): Promise<string> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "text",
        message: "Enter your Azure organisation's base Url: ",
        validate: (url: string) => validateUrl(url),
    });

    return promptData[PROMPT_NAME];
};

export const getAzureAccessToken = async (): Promise<string> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "text",
        message: "Enter your Azure personal access token: ",
        validate: (token: string) => validatePersonalAccessToken(token),
    });

    return promptData[PROMPT_NAME];
};

export const getGithubBaseUrl = async (): Promise<string> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "text",
        message: "Enter your Github organisation's base Url: ",
        validate: (url: string) => validateUrl(url),
    });

    return promptData[PROMPT_NAME];
};

export const getGithubAccessToken = async (): Promise<string> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "text",
        message: "Enter your Github personal access token: ",
        validate: (token: string) => validatePersonalAccessToken(token),
    });

    return promptData[PROMPT_NAME];
};

export const getGitlabBaseUrl = async (): Promise<string> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "text",
        message: "Enter your Gitlab organisation's base Url: ",
        validate: (url: string) => validateUrl(url),
    });

    return promptData[PROMPT_NAME];
};

export const getGitlabAccessToken = async (): Promise<string> => {
    const promptData: Answers<PromptType> = await prompts({
        name: PROMPT_NAME,
        type: "text",
        message: "Enter your Gitlab personal access token: ",
        validate: (token: string) => validatePersonalAccessToken(token),
    });

    return promptData[PROMPT_NAME];
};

const getEmptyConfig = (): Config => {
    return {
        startDate: new Date(),
        endDate: new Date(),
        azure: {
            enabled: false,
            baseUrl: "",
            personalAccessToken: "",
        },
        github: {
            enabled: false,
            baseUrl: "",
            personalAccessToken: "",
        },
        gitlab: {
            enabled: false,
            baseUrl: "",
            personalAccessToken: "",
        },
        httpTimeoutInMS: 5000,
    };
};

export const getConfigFromCli = async (): Promise<Config> => {
    const config: Config = getEmptyConfig();

    config.startDate = await getStartDate();
    config.endDate = await getEndDate(config.startDate);

    const platforms: string[] = await getPlatforms();
    config.azure.enabled = platforms.includes("Azure");
    config.github.enabled = platforms.includes("Github");
    config.gitlab.enabled = platforms.includes("Gitlab");

    if (config.azure.enabled) {
        config.azure.baseUrl = await getAzureBaseUrl();
        config.azure.personalAccessToken = await getAzureAccessToken();
    }

    if (config.github.enabled) {
        config.github.baseUrl = await getGithubBaseUrl();
        config.github.personalAccessToken = await getGithubAccessToken();
    }

    if (config.gitlab.enabled) {
        config.gitlab.baseUrl = await getGitlabBaseUrl();
        config.gitlab.personalAccessToken = await getGitlabAccessToken();
    }

    return config;
};
